### PR TITLE
Modulo-reduce msg32 inside RFC6979 nonce fn to match spec. Fixes #1063

### LINKED ISSUE #1063: ECDSA does not conform to RFC6979 for messages > curve_order
[RFC6979 3.2.d](https://datatracker.ietf.org/doc/html/rfc6979#section-3.2) says:

```
K = HMAC_K(V || 0x00 || int2octets(x) || bits2octets(h1))
```

where bits2octets is, as per [RFC6979 2.3.4](https://datatracker.ietf.org/doc/html/rfc6979#section-2.3.4) **curve order modulo-reduced** message:

```
   The bits2octets transform takes as input a sequence of blen bits and
   outputs a sequence of rlen bits.  It consists of the following steps:

   1.  The input sequence b is converted into an integer value z1
       through the bits2int transform:

          z1 = bits2int(b)
   2.  z1 is reduced modulo q, yielding z2 (an integer between 0 and
       q-1, inclusive):

          z2 = z1 mod q

       Note that since z1 is less than 2^qlen, that modular reduction
       can be implemented with a simple conditional subtraction:
       z2 = z1-q if that value is non-negative; otherwise, z2 = z1.

   3.  z2 is transformed into a sequence of octets (a sequence of rlen
       bits) by applying int2octets.
```

The implementation's `sign` takes `msg32` — not modulo-reduced `msg`, and passes it forward.

https://github.com/bitcoin-core/secp256k1/blob/0559fc6e41b65af6e52c32eb9b1286494412a162/src/secp256k1.c#L476

Seems like a bug, which does not exist in [go-btcec](https://github.com/btcsuite/btcd/tree/master/btcec) etc.

### PR DESCRIPTION
**Pull request by paulmillr**

### CODE REVIEW

**Review by real-or-random — COMMENTED** [review]
I think it's good to adhere to the spec.

If we want to change this, the right way will be to change it in `nonce_function_rfc6979` because the change you propose here will also affect callers who provide their custom nonce function. But that means we'll need to do another reduction in  `nonce_function_rfc6979`, change the nonce function API, or introduce strange  special casing. Hm, I don't know, not sure if it's worth the hassle.

Anyway, I think what we should do is to overhaul the ECDSA API (https://github.com/bitcoin-core/secp256k1/issues/974), and then we could provide a nice simple ECDSA function that uses a much faster nonce function than RFC6979.

**Comment by paulmillr** [comment]
Is it okay to cut a new release with the change (or an alternative similar change in nonce function)?

Before the refactoring, which could take a long time.

I develop secp256k1 in js and it seems like there are some assumptions in libraries that test against particular vectors e.g message > curve order. Releasing a new libsecp version would allow to have some "real" vectors to test against, since it's considered a reference implementation.

**Comment by real-or-random** [comment]
> Is it okay to cut a new release with the change (or an alternative similar change in nonce function)?

I'd be totally fine with adding a reduction to `nonce_function_rfc6979`. That's simple and the solution with the cleanest code (right now)  and while it makes signing slightly slowing, you anyway can't expect optimal performance from RFC6979 signing. How does that sound to others?

Well, releases are a separate issue... So far we never had a release (https://github.com/bitcoin-core/secp256k1/issues/286) but there's good progress recently (https://github.com/bitcoin-core/secp256k1/pull/1055).

**Comment by sipa** [comment]
I think we should:
* Make `nonce_function_rfc6979` do the modulo reduction, to comply with the spec.
* We could make `nonce_function_rfc6979` also reduce the output, so that it doesn't cause a second RNG invocation if the result exceed the order. This isn't an observable difference however, if RFC6979 is a PRF.
* Add a new fast, sane, nonce function with the simplest possible semantics around edge cases (something based on the BIP340 nonce functie, I imagine).
* Optionally, make this new nonce function the default, but leave the rfc6979 one in place.

**Comment by real-or-random** [comment]
> Add a new fast, sane, nonce function with the simplest possible semantics around edge cases (something based on the BIP340 nonce function, I imagine).

There's some discussion around this here: https://github.com/bitcoin-core/secp256k1/issues/757#issuecomment-644001840

File: src/secp256k1.c [hunk 1]
```diff
@@ -423,6 +423,10 @@ static int nonce_function_rfc6979(unsigned char *nonce32, const unsigned char *m
    unsigned int offset = 0;
    secp256k1_rfc6979_hmac_sha256 rng;
    unsigned int i;
+   secp256k1_scalar msg;
+   unsigned char msgmod32[32];
+   secp256k1_scalar_set_b32(&msg, msg32, NULL);
+   secp256k1_scalar_get_b32(msgmod32, &msg);
    /* We feed a byte array to the PRNG as input, consisting of:
     * - the private key (32 bytes) and message (32 bytes), see RFC 6979 3.2d.
```
Comment (real-or-random) [inline src/secp256k1.c:431]:
```suggestion
    * - the private key (32 bytes) and reduced message (32 bytes), see RFC 6979 3.2d.
```

**Review by real-or-random — COMMENTED** [review]
Looks good, can you squash?

**Comment by paulmillr** [comment]
@real-or-random done

**Review by real-or-random — APPROVED** [review]
ACK 45f37b650635e46865104f37baed26ef8d2cfb97

**Comment by siv2r** [comment]
ACK 45f37b6. The diff looks good. It reduces `msg32` to modulo curve order for rfc6979 nonce generation. All tests passed on my machine with `make check`.
<details>
<summary> make check result </summary>

```
make  check-am
make[1]: Entering directory '/home/siv2r/Projects/secp256k1'
make  check-TESTS
make[2]: Entering directory '/home/siv2r/Projects/secp256k1'
make[3]: Entering directory '/home/siv2r/Projects/secp256k1'
PASS: tests
PASS: exhaustive_tests
============================================================================
Testsuite summary for libsecp256k1 0.1.0-pre
============================================================================
# TOTAL: 2
# PASS:  2
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[3]: Leaving directory '/home/siv2r/Projects/secp256k1'
make[2]: Leaving directory '/home/siv2r/Projects/secp256k1'
make[1]: Leaving directory '/home/siv2r/Projects/secp256k1'
```

</details>

**Comment by sipa** [comment]
utACK 45f37b650635e46865104f37baed26ef8d2cfb97

**Comment by sipa** [comment]
Apparently this was detected to cause a divergence with the trezor firmware: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=43997

I haven't verified myself, but I assume it's because they use the unreduced msghash as well.

**Comment by prusnak** [comment]
> I haven't verified myself, but I assume it's because they use the unreduced msghash as well.

We will address this in the upcoming PR (issue https://github.com/trezor/trezor-firmware/issues/2085)

**Pull request merged**